### PR TITLE
[mono] Implement Arm intrinsics: ArmBase, Crc32

### PR DESCRIPTION
**Pull request by EgorBo**
Implement `System.Runtime.Intrinsics.Arm.ArmBase` and `*.Crc32` intrinsics for Arm64 with LLVM-backend. `ArmBase` is enabled by default, `Crc32` for LLVM-AOT can be enabled either via `mcpu=native` or `mattr=+crc` (e.g. when we crosscompile for iOS).

Asm: (`--aot=llvm,mcpu=native System.Private.CoreLib.dll` on Cortex-A57):
```asm
;; ArmBase

; int LeadingZeroCount(int value)
0000000000292c84 <System_Runtime_Intrinsics_Arm_ArmBase_LeadingZeroCount_int>:
  292c84:	clz	w0, w0
  292c88:	ret

; int LeadingZeroCount(uint value)
0000000000292c8c <System_Runtime_Intrinsics_Arm_ArmBase_LeadingZeroCount_uint>:
  292c8c:	clz	w0, w0
  292c90:	ret

; int ReverseElementBits(int value)
0000000000292c94 <System_Runtime_Intrinsics_Arm_ArmBase_ReverseElementBits_int>:
  292c94:	rbit	w0, w0
  292c98:	ret

; uint ReverseElementBits(uint value)
0000000000292c9c <System_Runtime_Intrinsics_Arm_ArmBase_ReverseElementBits_uint>:
  292c9c:	rbit	w0, w0
  292ca0:	ret

;; ArmBase.Arm64

; int LeadingSignCount(int value)
0000000000292cac <System_Runtime_Intrinsics_Arm_ArmBase_Arm64_LeadingSignCount_int>:
  292cac:	cls	w0, w0
  292cb0:	ret

; int LeadingSignCount(long value)
0000000000292cb4 <System_Runtime_Intrinsics_Arm_ArmBase_Arm64_LeadingSignCount_long>:
  292cb4:	cls	x0, x0
  292cb8:	ret

; int LeadingZeroCount(long value)
0000000000292cbc <System_Runtime_Intrinsics_Arm_ArmBase_Arm64_LeadingZeroCount_long>:
  292cbc:	clz	x0, x0
  292cc0:	ret

; int LeadingZeroCount(ulong value)
0000000000292cc4 <System_Runtime_Intrinsics_Arm_ArmBase_Arm64_LeadingZeroCount_ulong>:
  292cc4:	clz	x0, x0
  292cc8:	ret

; long ReverseElementBits(long value)
0000000000292ccc <System_Runtime_Intrinsics_Arm_ArmBase_Arm64_ReverseElementBits_long>:
  292ccc:	rbit	x0, x0
  292cd0:	ret

; ulong ReverseElementBits(ulong value)
0000000000292cd4 <System_Runtime_Intrinsics_Arm_ArmBase_Arm64_ReverseElementBits_ulong>:
  292cd4:	rbit	x0, x0
  292cd8:	ret

;; Crc32

; uint ComputeCrc32(uint crc, byte data)
0000000000292ce4 <System_Runtime_Intrinsics_Arm_Crc32_ComputeCrc32_uint_byte>:
  292ce4:	crc32b	w0, w0, w1
  292ce8:	ret

; uint ComputeCrc32(uint crc, ushort data)
0000000000292cec <System_Runtime_Intrinsics_Arm_Crc32_ComputeCrc32_uint_uint16>:
  292cec:	crc32h	w0, w0, w1
  292cf0:	ret

; uint ComputeCrc32(uint crc, uint data)
0000000000292cf4 <System_Runtime_Intrinsics_Arm_Crc32_ComputeCrc32_uint_uint>:
  292cf4:	crc32w	w0, w0, w1
  292cf8:	ret

; uint ComputeCrc32C(uint crc, byte data)
0000000000292cfc <System_Runtime_Intrinsics_Arm_Crc32_ComputeCrc32C_uint_byte>:
  292cfc:	crc32cb	w0, w0, w1
  292d00:	ret

; uint ComputeCrc32C(uint crc, ushort data)
0000000000292d04 <System_Runtime_Intrinsics_Arm_Crc32_ComputeCrc32C_uint_uint16>:
  292d04:	crc32ch	w0, w0, w1
  292d08:	ret

; uint ComputeCrc32C(uint crc, uint data)
0000000000292d0c <System_Runtime_Intrinsics_Arm_Crc32_ComputeCrc32C_uint_uint>:
  292d0c:	crc32cw	w0, w0, w1
  292d10:	ret

;; Crc32.Arm64

; uint ComputeCrc32(uint crc, ulong data)
0000000000292d1c <System_Runtime_Intrinsics_Arm_Crc32_Arm64_ComputeCrc32_uint_ulong>:
  292d1c:	crc32x	w0, w0, x1
  292d20:	ret

; uint ComputeCrc32C(uint crc, ulong data)
0000000000292d24 <System_Runtime_Intrinsics_Arm_Crc32_Arm64_ComputeCrc32C_uint_ulong>:
  292d24:	crc32cx	w0, w0, x1
  292d28:	ret
```
@tannergooding @echesakovMSFT Could you please take a quick look at asm ^ if it's ok or not.

`Vector64<T>`, `Vector128<T>`, `Vector<T>` are not enabled yet.

### CODE REVIEW

File: src/mono/mono/mini/local-propagation.c [hunk 1]
```diff
@@ -841,7 +841,7 @@ static gboolean
 mono_is_simd_accessor (MonoInst *ins)
 {
 	switch (ins->opcode) {
-#ifdef MONO_ARCH_SIMD_INTRINSICS
+#if defined(MONO_ARCH_SIMD_INTRINSICS) && !defined(TARGET_ARM64) // FIXME: enable for Arm64
 	case OP_INSERT_I1:
```
Comment (vargaz) [inline src/mono/mono/mini/local-propagation.c]: This needs to be fixed.

File: src/mono/mono/mini/mini-llvm.c [hunk 2]
```diff
@@ -8559,7 +8551,73 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
...
+			
+			LLVMValueRef args [2];
+			args [0] = add;
+			args [1] = LLVMConstInt (LLVMInt1Type (), 0, FALSE);
+			values [ins->dreg] = LLVMBuildCall (builder, get_intrins (ctx, ins->opcode == OP_LSCNT32 ? INTRINS_CTLZ_I32 : INTRINS_CTLZ_I64), args, 2, "");
+			break;
+		}
+
+		case OP_RBIT32:
+		case OP_RBIT64:
```
Comment (vargaz) [inline src/mono/mono/mini/mini-llvm.c]: All these could be a subset of the OP_XOP_ opcodes.

File: src/mono/mono/mini/mini-ops.h [hunk 3]
```diff
@@ -1534,3 +1534,15 @@ MINI_OP(OP_LZCNT32, "lzcnt32", IREG, IREG, NONE)
...
 MINI_OP(OP_POPCNT64, "popcnt64", LREG, LREG, NONE)
+
+#ifdef TARGET_ARM64
+MINI_OP(OP_LSCNT32, "lscnt32", IREG, IREG, NONE)
+MINI_OP(OP_LSCNT64, "lscnt64", LREG, LREG, NONE)
+MINI_OP(OP_RBIT32, "rbit32", IREG, IREG, NONE)
+MINI_OP(OP_RBIT64, "rbit64", LREG, LREG, NONE)
+
+MINI_OP(OP_AARCH64_CRC32C, "aarch64_crc32c", IREG, IREG, IREG)
+MINI_OP(OP_AARCH64_CRC32CX, "aarch64_crc32cd", IREG, IREG, LREG)
```
Comment (vargaz) [inline src/mono/mono/mini/mini-ops.h]: We use ARM64 as the arch name.

File: src/mono/mono/mini/simd-intrinsics-netcore.c [hunk 4]
```diff
@@ -291,6 +266,34 @@ get_vector_t_elem_type (MonoType *vector_type)
 	return etype;
 }
 
+// FIXME: implement for Arm64
+#ifndef TARGET_ARM64
+
```
Comment (vargaz) [inline src/mono/mono/mini/simd-intrinsics-netcore.c]: Would be better to add a new define for this, or use #ifdef AMD64.

File: src/mono/mono/mini/mini-ops.h (hunk 3 above)
Comment (EgorBo) [reply src/mono/mono/mini/mini-ops.h]: Ah, sure, I just copied the original LLVM intrinsics' names: `@llvm.aarch64.crc*`.

File: src/mono/mono/mini/local-propagation.c (hunk 1 above)
Comment (vargaz) [reply src/mono/mono/mini/local-propagation.c]: It might be a good idea to reserve ARCH_SIMD_INTRINSICS for the existing code, and just unconditionally add the new code to simd-intrinsics-netcore.c, the define doesn't make too much sense any more.

File: src/mono/mono/mini/mini.c [hunk 5]
```diff
@@ -4345,6 +4345,11 @@ mini_get_cpu_features (MonoCompile* cfg)
 	}
 #endif
 
+#if defined(TARGET_ARM64)
+	// All Arm64 devices have this set
+	features |= MONO_CPU_ARM64_BASE; 
+#endif
```
Comment (vargaz) [inline src/mono/mono/mini/mini.c]: Why is this needed if its always set ?

Comment (tannergooding) [reply src/mono/mono/mini/mini.c]: At least for CoreCLR, we support disabling this via `COMPlus_EnableHWIntrinsic=0` so that you can easily test software fallbacks, etc.

**Comment by tannergooding** [comment]
> ArmBase is enabled by default, Crc32 for LLVM-AOT can be enabled either via mcpu=native or mattr=+crc (e.g. when we crosscompile for iOS).

Is `mattr=+crc` something from LLVM or something from Mono? If the latter, is it worth making this consistent with the names we were going to use for crossgen? (CC. @davidwrighton)

File: src/mono/mono/mini/mini.c (hunk 5 above)
Comment (EgorBo) [reply src/mono/mono/mini/mini.c]: Yes, it makes it possible to disable ArmBase via `mattr=-base` if needed

**Comment by EgorBo** [comment]
> > ArmBase is enabled by default, Crc32 for LLVM-AOT can be enabled either via mcpu=native or mattr=+crc (e.g. when we crosscompile for iOS).
> 
> Is `mattr=+crc` something from LLVM or something from Mono? If the latter, is it worth making this consistent with the names we were going to use for crossgen? (CC. @davidwrighton)

It's mono, but it's based on LLVM naming convention for features (see `llvm::sys::getHostCPUFeatures`)
However, we also pass it as is to LLVM's `opt` and `llc` tools.

**Comment by EgorBo** [comment]
@vargaz could you please re-review?

File: src/mono/mono/mini/aot-compiler.c [hunk 6]
```diff
@@ -8111,7 +8111,14 @@ parse_cpu_features (const gchar *attr)
 		feature = (MonoCPUFeatures) (MONO_CPU_X86_FULL_SSEAVX_COMBINED & ~feature);
 	
 #elif defined(TARGET_ARM64)
-	// TODO: neon, sha1, sha2, asimd, etc...
+	if (!strcmp (attr + prefix, "base"))
+		feature = MONO_CPU_ARM64_BASE;
+	else if (!strcmp (attr + prefix, "neon"))
+		feature = MONO_CPU_ARM64_NEON;
+	else if (!strcmp (attr + prefix, "crypto"))
```
Comment (davidwrighton) [inline src/mono/mono/mini/aot-compiler.c]: Instead of using the crypto name here, could we use "aes" and "sha2" as those are the names that match up to individual instruction set usage on llvm.

**Comment by EgorBo** [comment]
@tannergooding I've just tried the lates daily .NET Core SDK (5.0.100-preview.3.20181.13) with System.Runtime.Intrinsics.Experimental (5.0.0-preview.3.20181.9) on my Cortex A57:
```
ArmBase.IsSupported: True
ArmBase.Arm64.IsSupported: True
AdvSimd.IsSupported: True
AdvSimd.Arm64.IsSupported: True
Aes.IsSupported: True
Aes.Arm64.IsSupported: True
Crc32.IsSupported: True
Crc32.Arm64.IsSupported: False
Sha1.IsSupported: True
Sha1.Arm64.IsSupported: True
Sha256.IsSupported: True
Sha256.Arm64.IsSupported: True
```
Any idea why `Crc32.Arm64.IsSupported` is only thing to return False here ? (arm64 arch).

**Comment by tannergooding** [comment]
> Any idea why Crc32.Arm64.IsSupported is only thing to return False here ? (arm64 arch).

@davidwrighton, possibly something missed in the fixes that just went in?

File: src/mono/mono/mini/aot-compiler.c (hunk 6 above)
Comment (EgorBo) [reply src/mono/mono/mini/aot-compiler.c:8118]: Here is what LLVM (getHostCPUFeatures) returns for my CPU: X: `neon fp-armv8 crypto crc`
Here is what /proc/cpuinfo says: `fp asimd evtstrm aes pmull sha1 sha2 crc32 cpuid`

**Comment by tannergooding** [comment]
@EgorBo, looks like it probably doesn't have some of the fixes that went in recently.

The SDK reports the following, which is 13 days old (https://github.com/dotnet/runtime/commit/bc28cbef85):
```
Host (useful for support):
  Version: 5.0.0-preview.3.20169.1
  Commit:  bc28cbef85
```

There were some important fixes (such as https://github.com/dotnet/runtime/pull/34139) that only went in about 5 days ago.

**Comment by EgorBo** [comment]
@tannergooding thanks, forgot that daily sdks are not entirely daily :) 

I noticed an API problem here it took me a while to understand.
So imagine my hw doesn't support Sha256.

`Sha256.IsSupported` returns `false` (expected)
`Sha256.Arm64.IsSupported` returns `true` (unexpected) - because Sha256 doesn't define a separate Arm64 impl it returns ArmBase for Arm64.

**Comment by echesakovMSFT** [comment]
> I noticed an API problem here it took me a while to understand.
> So imagine my hw doesn't support Sha256.
> Sha256.IsSupported returns false (expected)
> Sha256.Arm64.IsSupported returns true (unexpected) - because Sha256 doesn't define a separate Arm64 impl it returns ArmBase.Arm64.IsSupported for my Sha256.Arm64.IsSupported call.

@EgorBo  This sounds as the issue I faced recently (see my comment https://github.com/dotnet/runtime/pull/34107#issue-393925654). Should be fixed now.

File: src/mono/mono/mini/aot-compiler.c (hunk 6 above)
Comment (davidwrighton) [reply src/mono/mono/mini/aot-compiler.c:8118]: Which version of LLVM are you working with? I believe this has changed in more recent versions of llvm, and is a total mess. As I've been finishing up my related work with crossgen2 in pr #33274, I've noticed that there doesn't seem to be a compiler switch that enables only sha1, only sha2 which enables both sha1 and sha256. In addition, the behavior of the crypto switch is that it varies based on which baseline architecture is in use. So I believe crypto,armv8.1a has a different meaning than crypto,v8.3a, with regards to which crypto instructions are enabled which is extremely confusing. I believe we should try to make sure these parameters have a direct match to the CPU feature flags, and if that means we need to use different parameter names than llvm, so be it. Looking at the output of `llc-9 -march=aarch64 -mcpu=help`, I see the following relevant features available for the aarch64 architecture.

```
lse                                - Enable ARMv8.1 Large System Extension (LSE) atomic instructions.
neon                               - Enable Advanced SIMD instructions.
sha2                               - Enable SHA1 and SHA256 support.
crc                                - Enable ARMv8 CRC-32 checksum instructions.
```

My proposal is that we match llvm for the `lse`, `neon`, `sha2`, and `crc` parameters, except that we also add another `sha1` which represents access to the `Sha1` managed apis, and `base` which represents access to the `ArmBase` set of managed apis.

**Comment by EgorBo** [comment]
> > I noticed an API problem here it took me a while to understand.
> > So imagine my hw doesn't support Sha256.
> > Sha256.IsSupported returns false (expected)
> > Sha256.Arm64.IsSupported returns true (unexpected) - because Sha256 doesn't define a separate Arm64 impl it returns ArmBase.Arm64.IsSupported for my Sha256.Arm64.IsSupported call.
> 
> @EgorBo This sounds as the issue I faced recently (see my comment [#34107 (comment)](https://github.com/dotnet/runtime/pull/34107#issue-393925654)). Should be fixed now.

um.. I don't see how it can be implemented in Jit, here is what I mean: https://sharplab.io/#v2:EYLgHgbALANALiAlgGwD4AEBMBGAsAKAKP3QGYACLcgQQCcBbAIQEMBnAUwIG8Dy/KK6bBHLAA9mOTkAkqwDKAVwAOSsbTjsAJuQC8APnJxaC9gG5e/C3zKVMNBtCvke+fm4GVhoiVNmKVahra+obGZk4AvgRRhCSCdnIAFsyYAKwiIPZMbJz4Lm42QiLikjLyyqrqWroGRibmrpaNfAD0LeRJKenkmmLsrAB2AORw5MkAbuzkiHCs5GIA7gNZjvgxBIV2QgDs3E5tHWLTo8gSANZzyIhnU9LkAMaJ7Pdn5ABmah3JaSLM48woZjAFAzACeY3YtFybgOwAUo0Qy2kABljkM5gAie7MZBSEpSOjZDgtQnQEAgADm7DgAH0/BVAloABQASgxTkKXnx5AAYhJWTUvl0IAA6UlQEX0gJVTQNCJAA

**Comment by echesakovMSFT** [comment]
> So I'd add an empty Arm64 nested class with the only property - IsSupported (to return what Sha256.IsSupported returns)

@EgorBo  Ah okay - thanks for explanation! I mixed this up with another issue. I guess we could workaround this by defining Sha256.Arm64 class that contains only one property IsSupported always returning false but this just doesn't sound right to me.

Perhaps deriving Sha256 class from ArmBase wasn't a good idea.

**Comment by tannergooding** [comment]
Yes, we would need to have an explicit class that plays into the normal lookup checks.

This is probably something that should have an issue and should go through API review.

**Review by vargaz — APPROVED** [review]

File: src/mono/mono/mini/aot-compiler.c (hunk 6 above)
Comment (EgorBo) [reply src/mono/mono/mini/aot-compiler.c]: Got it, remove "crypto" from the PR and left only "crc" since it's the only set my PR implements (+ base one).
We use LLVM 6 (and migrating to LLVM 9)
